Optimize loop of phi_function

romainbrenguier opened the pull request:
This function was performing unnecessary lookups (even after `ranget`), we avoid this by introducing a function which goes through both maps and use the fact that they are ordered.

I tested this optimization on 73 methods from tika which where taking between 600ms and 30s with --show-vcc and --verbosity 0,  the total time was improved by 20% (from 229s to 185s). Here is a comparison of the execution times 
![perf_out](https://user-images.githubusercontent.com/11854194/49379661-8b60aa80-f707-11e8-8800-08b4862ce8e0.png)




<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [na] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

romainbrenguier committed:
Remove unused range

romainbrenguier committed:
Return early if nothing to do in phi_function

romainbrenguier committed:
Use a for_each2 function iterating over two maps

Inside phi_function we can avoid a lot of lookups if we iterate over the
two maps at the same time, since they are ordered.

tautschnig commented on src/goto-symex/symex_goto.cpp:
```diff
@@ -353,138 +353,155 @@ void goto_symext::merge_value_sets(
+/// second one then j is 0, and when the first map has no entry for k then i = 0
+static void for_each2(
+  std::map<irep_idt, std::pair<ssa_exprt, unsigned>> first_map,
+  std::map<irep_idt, std::pair<ssa_exprt, unsigned>> second_map,
```
Why are these maps passed by value?

romainbrenguier replied:
Yes it's a mistake

tautschnig commented on src/goto-symex/symex_goto.cpp:
```diff
@@ -353,138 +353,155 @@ void goto_symext::merge_value_sets(
+        const ssa_exprt &ssa,
+        unsigned goto_count,
+        unsigned dest_count)
+    {
```
As a first refactoring, and without making any other changes, can we please move this loop body in a function of its own? This will make all other changes a lot easier to read.

tautschnig reviewed (changes requested):
The performance improvement is of course desirable, but could the changes please be done in a way so that the diff (when switching on skip-whitespace) is readable?

romainbrenguier commented:
I was passing the maps by copy instead of reference, when correcting that the performance improvement is now of 33%
![perf_out](https://user-images.githubusercontent.com/11854194/49404221-bc140480-f746-11e8-8f0a-b985e96c4f23.png)

romainbrenguier committed:
Extract function from the loop in phi_function

This simplifies phi_function which was too long.

romainbrenguier committed:
Compute goto_count and dest_count before merge_names

Since these are used several times in the function we can avoid some
lookups by pre-computing it.
We make them parameters to allow later commits to give them without
lookup.

romainbrenguier committed:
Document merge_names function

romainbrenguier committed:
Inline do_simplify

This will allow to more easily extract the code in the loop into a new
function.

romainbrenguier commented:
@tautschnig this is ready for re-review

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: c79694223b989e27a5f9802a310d1dac4f3f37c0).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/93674773

tautschnig commented on src/goto-symex/symex_goto.cpp:
```diff
@@ -377,13 +377,12 @@ void goto_symext::phi_function(
-  {
-    diff_guard=goto_state.guard;
+  if(all_current_names_range.empty())
+    return;
```
Ok, but then you don't need the declaration above either. Could you make the first definition below also be the declaration?

tautschnig commented on src/goto-symex/symex_goto.cpp:
```diff
@@ -388,104 +499,18 @@ void goto_symext::phi_function(
+      symex_config.simplify_opt,
+      target,
+      l1_identifier,
+      ssa);
```
If we pass in `ssa` anyway, then `l1_identifier` should be computed from that in the body of `merge_names`.

tautschnig commented on src/goto-symex/symex_goto.cpp:
```diff
@@ -354,6 +354,117 @@ void goto_symext::merge_value_sets(
+  // once the thread is executed)
+  if(
+    !ssa.get_level_0().empty() &&
+      ssa.get_level_0() != std::to_string(dest_state.source.thread_nr))
```
Would you mind adding an extra commit that avoids this duplicate call to `get_level_0()`?

romainbrenguier replied:
done

tautschnig reviewed (approved):
A couple of suggestions left, I think we should be good to go once these are in.

kroening commented:
The clang-format commit should be merged into the relevant earlier commits.

romainbrenguier committed:
Avoid duplicate get_level_0 call

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: bbec003e8547440ee51931bb5b2c52fab1479737).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/93783524
